src/common/buffer_seastar.cc

```diff
@@ -24,9 +27,25 @@ namespace ceph::buffer {
 
 class raw_seastar_foreign_ptr : public raw {
   seastar::foreign_ptr<temporary_buffer> ptr;
+  seastar::alien::instance& alien;
  public:
   raw_seastar_foreign_ptr(temporary_buffer&& buf)
-    : raw(buf.get_write(), buf.size()), ptr(std::move(buf)) {}
+    : raw(buf.get_write(), buf.size()), ptr(std::move(buf)),
+      alien(seastar::engine().alien()) {}
+
+  ~raw_seastar_foreign_ptr() {
+    if (ptr.get_owner_shard() != seastar::this_shard_id()) {
+      char buf[200] = {0};
+      assert(pthread_getname_np(pthread_self(), buf ,200) == 0);
+      if (std::string_view str(buf);
+          str == "alien-store-tp" || str == "bstore_kv_final") {
+        // we should let a seastar reactor destroy this memory, we are alien
+        seastar::alien::submit_to(alien, ptr.get_owner_shard(), [this] () noexcept {
+          return ptr.destroy();
```

Comment:
destroy() calls destroy_on which nullifies the wrapped pointer.
```
    static future<> destroy_on(PtrType p, unsigned cpu) noexcept {
        if (p) {
            if (cpu != this_shard_id()) {
                return smp::submit_to(cpu, [v = std::move(p)] () mutable {
                    // Destroy the contained pointer. We do this explicitly
                    // in the current shard, because the lambda is destroyed
                    // in the shard that submitted the task.
                    v = {};
                });
            } else {
                p = {};
            }
        }
        return make_ready_future<>();
    }
```